pkg/driver/driver_aws.go

```diff
@@ -115,30 +120,149 @@ func (d *AWSDriver) Create() (string, string, error) {
 		networkInterfaceSpecs = append(networkInterfaceSpecs, spec)
 	}
 
-	// Specify the details of the machine
-	inputConfig := ec2.RunInstancesInput{
-		ImageId:           aws.String(d.AWSMachineClass.Spec.AMI),
-		InstanceType:      aws.String(d.AWSMachineClass.Spec.MachineType),
-		MinCount:          aws.Int64(1),
-		MaxCount:          aws.Int64(1),
-		UserData:          &UserDataEnc,
-		KeyName:           aws.String(d.AWSMachineClass.Spec.KeyName),
-		NetworkInterfaces: networkInterfaceSpecs,
-		IamInstanceProfile: &ec2.IamInstanceProfileSpecification{
-			Name: &(d.AWSMachineClass.Spec.IAM.Name),
-		},
-		BlockDeviceMappings: blkDeviceMappings,
-		TagSpecifications:   []*ec2.TagSpecification{tagInstance, tagVolume},
-	}
+	var instance *ec2.Instance
+	if d.AWSMachineClass.Spec.SpotPrice == nil {
+		// Specify the details of the machine
+		inputConfig := ec2.RunInstancesInput{
+			ImageId:           aws.String(d.AWSMachineClass.Spec.AMI),
+			InstanceType:      aws.String(d.AWSMachineClass.Spec.MachineType),
+			MinCount:          aws.Int64(1),
+			MaxCount:          aws.Int64(1),
+			UserData:          &UserDataEnc,
+			KeyName:           aws.String(d.AWSMachineClass.Spec.KeyName),
+			NetworkInterfaces: networkInterfaceSpecs,
+			IamInstanceProfile: &ec2.IamInstanceProfileSpecification{
+				Name: &(d.AWSMachineClass.Spec.IAM.Name),
+			},
+			BlockDeviceMappings: blkDeviceMappings,
+			TagSpecifications:   []*ec2.TagSpecification{tagInstance, tagVolume},
+		}
 
-	runResult, err := svc.RunInstances(&inputConfig)
-	if err != nil {
-		metrics.APIFailedRequestCount.With(prometheus.Labels{"provider": "aws", "service": "ecs"}).Inc()
-		return "Error", "Error", err
+		runResult, err := svc.RunInstances(&inputConfig)
+		if err != nil {
+			metrics.APIFailedRequestCount.With(prometheus.Labels{"provider": "aws", "service": "ecs"}).Inc()
+			return "Error", "Error", err
+		}
+		metrics.APIRequestCount.With(prometheus.Labels{"provider": "aws", "service": "ecs"}).Inc()
+
+		instance = runResult.Instances[0]
+	} else {
+		var clusterName string
+		for key := range d.AWSMachineClass.Spec.Tags {
+			if strings.Contains(key, "kubernetes.io/cluster/") {
+				clusterName = key
+			}
+		}
+
+		spotPrice := d.AWSMachineClass.Spec.SpotPrice
+		if len(*spotPrice) == 0 {
+			spotPrice = nil
+		}
+
+		spotInstanceRequestInput := &ec2.RequestSpotInstancesInput{
```

Comment:
did you consider using spot-fleet one-time request rather than spotRequest? I'm not sure if makes much of a difference with the current implementation since it only accepts 1 instance type anyways. But if multiple instance types could be passed, spot-fleet one-time request could be used with the capacity-optimized spot allocation strategy to select the best instance type to use. https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-fleet-requests.html

Reply:
We have a machine-object which I believe maps to the SpotRequest of a single instance here. That could block the natural use of the SpotFleet. @zuzzas do you think there could be a way to get the SpotFleet in with the current machine-model?

ClusterAPI recently came up with the MachinePool abstraction: https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20190919-machinepool-api.md . I am curious if we could also adopt something similar if that helps us bring the SpotFleet in. 
- Another prominent benefit then would be for Azure, we can then directly consume VMSS. [@MSSedusch ] .

If we agree overall, we could open a design-doc to discuss MachinePool+SpotFleet for MCM. cc @rewiko @zuzzas @prashanth26 @bwagner5

Reply:
Yes, I was thinking for the current machine-model. Spot Fleet has request types of "persistent" or "one-time".  The "persistent" type will maintain the desired capacity similar to how autoscaling groups work. The "one-time" type will launch the instances at the desired capacity, but won't maintain the capacity. So a one-time spot fleet request with a desired capacity of 1 would mirror the functionality that RequestSpotInstances provides but would also allow for the use of SpotAllocationStrategies. The capacity-optimized spot allocation strategy allows Spot Fleet to accept multiple instance-types and then it will choose one with a lot of capacity thereby reducing the chance of a spot interruption. 

As far as machine-pool is concerned, ASG should take care of all of the use-cases even with spot, without the need to pull in spot fleet.

Reply:
As a third option, we recently implemented spot instances within the openshift cluster-api provider for AWS by leveraging the `RunInstancesInput`. I think we have the same effect as the current implementation but AWS handles a lot of it for us by setting up the spot instance request and checking it for us. Don't you know if you considered this at all?

You may be able to simplify this a lot if you followed the same route as us, create the `InstanceMarketOptions` as below, add that to `RunInstancesInput`, and then if the instance can't be fulfilled immediately, AWS returns a 400 so you can tell that the creation failed, otherwise it accepts it and you get a ressponse like a normal `RunInstances`.
https://github.com/JoelSpeed/cluster-api-provider-aws/blob/e1b2632853623790c11d60f1bf2a34cc7d697eec/pkg/actuators/machine/instances.go#L379-L407

Behind the scenes it relies on an individual spot instance request as you have implemented here.

Reply:
@JoelSpeed, well, that's embarrassing. I've completely glosser over this attribute in `RunInstancesInput`. I'll refactor my PR today, should shave off a lot of useless code.

Thank you!

Reply:
No worries, happy to help 😅

Reply:
@zuzzas fwiw you can get more context on our reasoning and choices (which seems pretty aligned with this PR) here https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/spot-instances.md

Reply:
Thanks to colleagues from Red Hat, I've significantly simplified this PR.